File: app/src/main/java/com/quigglesproductions/secureimageviewer/ui/enhancedfolderviewer/EnhancedFolderViewerFragment.java
package com.quigglesproductions.secureimageviewer.ui.enhancedfolderviewer;

import static androidx.navigation.Navigation.findNavController;

import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.ContextMenu;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.widget.AdapterView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavDirections;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.android.material.snackbar.Snackbar;
import com.google.gson.reflect.TypeToken;
import com.mikelau.views.shimmer.ShimmerRecyclerViewX;
import com.quigglesproductions.secureimageviewer.R;
import com.quigglesproductions.secureimageviewer.SortType;
import com.quigglesproductions.secureimageviewer.database.enhanced.EnhancedDatabaseHandler;
import com.quigglesproductions.secureimageviewer.databinding.FragmentFolderViewBinding;
import com.quigglesproductions.secureimageviewer.managers.ApplicationPreferenceManager;
import com.quigglesproductions.secureimageviewer.managers.FolderManager;
import com.quigglesproductions.secureimageviewer.managers.NotificationManager;
import com.quigglesproductions.secureimageviewer.models.enhanced.datasource.IFileDataSource;
import com.quigglesproductions.secureimageviewer.models.enhanced.datasource.IFolderDataSource;
import com.quigglesproductions.secureimageviewer.models.enhanced.file.EnhancedDatabaseFile;
import com.quigglesproductions.secureimageviewer.models.enhanced.file.EnhancedOnlineFile;
import com.quigglesproductions.secureimageviewer.models.enhanced.file.IDisplayFile;
import com.quigglesproductions.secureimageviewer.models.enhanced.folder.EnhancedDatabaseFolder;
import com.quigglesproductions.secureimageviewer.models.enhanced.folder.IDisplayFolder;
import com.quigglesproductions.secureimageviewer.models.enhanced.metadata.IFileMetadata;
import com.quigglesproductions.secureimageviewer.room.databases.file.relations.FileWithMetadata;
import com.quigglesproductions.secureimageviewer.room.databases.file.relations.FolderWithFiles;
import com.quigglesproductions.secureimageviewer.ui.EnhancedMainMenuActivity;
import com.quigglesproductions.secureimageviewer.ui.EnhancedMainMenuViewModel;
import com.quigglesproductions.secureimageviewer.ui.SecureFragment;
import com.techyourchance.threadposter.BackgroundThreadPoster;
import com.techyourchance.threadposter.UiThreadPoster;

import java.lang.reflect.Type;
import java.net.MalformedURLException;
import java.util.ArrayList;
import java.util.List;

public class EnhancedFolderViewerFragment extends SecureFragment {
    private static final int CONTEXTMENU_INFO = 0;
    private static final int CONTEXTMENU_SET_THUMBNAIL = 1;
    private static final int CONTEXTMENU_UPLOAD = 2;

    private static final int LIST_UPDATE_TRIGGER_THRESHOLD = 33;
    EnhancedFileGridRecyclerAdapter enhancedAdapter;
    //GridView gridview;
    FragmentFolderViewBinding binding;
    IDisplayFolder selectedFolder;
    private boolean scrollBottomReached;
    SortType startingSortType;
    EnhancedFolderViewerViewModel viewModel;
    ShimmerRecyclerViewX recyclerView;
    private final BackgroundThreadPoster backgroundThreadPoster = new BackgroundThreadPoster();
    private final UiThreadPoster uiThreadPoster = new UiThreadPoster();
    public EnhancedFolderViewerFragment(){
    }
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        //postponeEnterTransition();
        setHasOptionsMenu(true);
        binding = FragmentFolderViewBinding.inflate(inflater,container,false);
        View root = binding.getRoot();
        viewModel = new ViewModelProvider(this).get(EnhancedFolderViewerViewModel.class);
        EnhancedMainMenuViewModel mainMenuViewModel = new ViewModelProvider(this).get(EnhancedMainMenuViewModel.class);
        //gridview = binding.fileGridview;
        recyclerView = binding.fileShimmerRecyclerView;
        enhancedAdapter = new EnhancedFileGridRecyclerAdapter(getContext());
        enhancedAdapter.setStateRestorationPolicy(RecyclerView.Adapter.StateRestorationPolicy.PREVENT_WHEN_EMPTY);
        int columnCount = getResources().getInteger(R.integer.column_count_filelist);
        final GridLayoutManager layoutManager = new GridLayoutManager(getContext(),columnCount);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setAdapter(enhancedAdapter);
        recyclerView.showShimmerAdapter();

        restoreInstanceState(savedInstanceState);
        /*gridview.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int position, long l) {
                NavDirections action = EnhancedFolderViewerFragmentDirections.actionEnhancedFolderViewerFragmentToEnhancedFileViewFragment(position);
                Navigation.findNavController(view).navigate(action);
                //Intent intent = new Intent(getContext(), EnhancedFileViewActivity.class);
                //intent.putExtra("position",position);
                //startActivity(intent);
            }
        });*/
        enhancedAdapter.setOnClickListener(new EnhancedFileGridRecyclerAdapter.EnhancedRecyclerViewOnClickListener() {
            @Override
            public void onClick(int position) {
                saveInstanceState();
                NavDirections action = EnhancedFolderViewerFragmentDirections.actionEnhancedFolderViewerFragmentToEnhancedFileViewFragment(position);
                findNavController(root).navigate(action);
            }

            @Override
            public void onCreateContextMenu(ContextMenu menu, View v, ContextMenu.ContextMenuInfo menuInfo) {
                menu.setHeaderTitle("Options");
                AdapterView.AdapterContextMenuInfo cmi = (AdapterView.AdapterContextMenuInfo) menuInfo;
                //EnhancedFile selectedFile = enhancedAdapter.get(position);
                IDisplayFile selectedFile = enhancedAdapter.get(cmi.position);
                menu.add(CONTEXTMENU_INFO, cmi.position, 0, "Info");
                if(selectedFolder.getFolderOrigin() == FolderOrigin.LOCAL)
                    menu.add(CONTEXTMENU_SET_THUMBNAIL, cmi.position, 0, "Set as Thumbnail");
            }
        });
        //gridview.setAdapter(adapter);
        selectedFolder = FolderManager.getInstance().getCurrentFolder();

        viewModel.getFiles().observe(getViewLifecycleOwner(), new Observer<ArrayList<IDisplayFile>>() {
            @Override
            public void onChanged(ArrayList<IDisplayFile> enhancedFiles) {
                enhancedAdapter.setFiles(enhancedFiles);
                container.getViewTreeObserver().addOnPreDrawListener(new ViewTreeObserver.OnPreDrawListener() {
                    @Override
                    public boolean onPreDraw() {
                        container.getViewTreeObserver().removeOnPreDrawListener(this);
                        //startPostponedEnterTransition();
                        return true;
                    }
                });
            }
        });
        viewModel.getFiles().observe(getViewLifecycleOwner(),enhancedAdapter::setFiles);
        registerForContextMenu(recyclerView);
        switch (selectedFolder.getFolderOrigin()){
            case ONLINE:
                startingSortType = ApplicationPreferenceManager.getInstance().getOnlineFolderSortType();
                break;
            case LOCAL:
            case ROOM:
                startingSortType = ApplicationPreferenceManager.getInstance().getOfflineFolderSortType();
                break;
            default:
                startingSortType = SortType.NAME_ASC;
        }
        /*recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                //int test = layoutManager.findLastCompletelyVisibleItemPosition();
                //int test2 = layoutManager.findLastVisibleItemPosition();
                int total = enhancedAdapter.getItemCount();
                //int test3 = total - (columnCount % total);
                if(layoutManager.findLastCompletelyVisibleItemPosition() >total-LIST_UPDATE_TRIGGER_THRESHOLD &&layoutManager.findLastCompletelyVisibleItemPosition() <=total-1){
                    if(scrollBottomReached == false) {
                        scrollBottomReached = true;
                        if(selectedFolder.getDataSource().moreItemsAvailable()){
                            try {
                                selectedFolder.getDataSource().getFilesFromDataSource(getContext(),new IFolderDataSource.FolderDataSourceCallback() {
                                    @Override
                                    public void FolderFilesRetrieved(List<EnhancedFile> files, Exception exception) {
                                        enhancedAdapter.addFiles(files);
                                        scrollBottomReached = false;
                                    }
                                },startingSortType);
                            } catch (MalformedURLException e) {
                                throw new RuntimeException(e);
                            }
                        }
                        else
                            scrollBottomReached = true;
                    }
                }
            }

            /*@Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
                if(!recyclerView.canScrollVertically(1)){

                }
            }*//*
        });*/


        return root;
    }

    private void restoreInstanceState(Bundle savedInstanceState){
        //if(savedInstanceState != null) {
        String fileListJson = viewModel.getState().get("FileList");
            //String fileListJson = savedInstanceState.getString("FileList");
        Type listType;
            if (fileListJson != null) {
                switch (selectedFolder.getFolderOrigin()){
                    case ONLINE:
                        listType = new TypeToken<ArrayList<EnhancedOnlineFile>>() {
                        }.getType();
                        break;
                    case LOCAL:
                        listType = new TypeToken<ArrayList<EnhancedDatabaseFile>>() {
                        }.getType();
                        break;
                    case ROOM:
                        listType = new TypeToken<ArrayList<FileWithMetadata>>() {
                        }.getType();
                        break;
                    default:
                        listType = null;
                }

                ArrayList<IDisplayFile> files = getGson().fromJson(fileListJson, listType);
                enhancedAdapter.setFiles(files);
                recyclerView.hideShimmerAdapter();
            }
        //}
    }

    private void saveInstanceState(){
        if(enhancedAdapter.getItemCount()>0)
            viewModel.getState().set("FileList",getGson().toJson(enhancedAdapter.getFiles()));
    }

    @Override
    public boolean onContextItemSelected(@NonNull MenuItem item) {
        IDisplayFile selectedFile = enhancedAdapter.get(item.getItemId());
        switch (item.getGroupId()){
            case CONTEXTMENU_INFO:
                //new ItemInfoDialog(adapter.getItem(item.getItemId())).show(getSupportFragmentManager(),ItemInfoDialog.TAG);
                BottomSheetDialog bottomSheetDialog = new BottomSheetDialog(getContext());
                bottomSheetDialog.setContentView(R.layout.bottomdialog_fileinfo);

                TextView itemNameText = bottomSheetDialog.findViewById(R.id.item_name);
                TextView folderNameText = bottomSheetDialog.findViewById(R.id.folder_name);
                TextView artistNameText = bottomSheetDialog.findViewById(R.id.artist_name);
                TextView catagoriesText = bottomSheetDialog.findViewById(R.id.catagories);
                TextView subjectsText = bottomSheetDialog.findViewById(R.id.subjects);
                selectedFile.getDataSource().getFileMetadata(requiresRequestManager(),new IFileDataSource.DataSourceFileMetadataCallback() {
                    @Override
                    public void FileMetadataRetrieved(IFileMetadata metadata, Exception exception) {
                        //selectedFile.metadata = metadata;
                        itemNameText.setText(selectedFile.getName());
                        folderNameText.setText(selectedFolder.getName());
                        artistNameText.setText(selectedFile.getArtistName());
                        catagoriesText.setText(selectedFile.getCatagoryListString());
                        subjectsText.setText(selectedFile.getSubjectListString());
                        bottomSheetDialog.create();
                        bottomSheetDialog.show();
                    }
                });

                break;
            case CONTEXTMENU_SET_THUMBNAIL:
                IDisplayFile file = enhancedAdapter.get(item.getItemId());
                if(file instanceof FileWithMetadata) {
                    backgroundThreadPoster.post(()->{
                        getFileDatabase().folderDao().setThumbnail((FolderWithFiles)selectedFolder,(FileWithMetadata) file);
                    });

                }
                break;
            case CONTEXTMENU_UPLOAD:
                /*AuthManager.getInstance().performActionWithFreshTokens(context, new AuthState.AuthStateAction() {
                    @Override
                    public void execute(@Nullable String accessToken, @Nullable String idToken, @Nullable AuthorizationException ex) {
                        RequestManager.getInstance().getRequestService().uploadFile(accessToken, adapter.getItem(item.getItemId()), new RequestManager.RequestResultCallback<EnhancedDatabaseFile, Exception>() {
                            @Override
                            public void RequestResultRetrieved(EnhancedDatabaseFile result, Exception exception) {
                                result.getIsUploaded();
                            }
                        });
                    }
                });*/

                break;
        }
        return true;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setTitle(selectedFolder.getName());
        restoreInstanceState(savedInstanceState);
        getFolderFiles(getContext(),viewModel);

    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.menu_folderview_fragment, menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()){
            case R.id.folderview_fragment_sort:
                showSortDialog();
                break;
            default:
                return false;
        }
        return true;
    }

    private void showSortDialog(){
        final CharSequence[] items = {"Name A-Z", "Name Z-A", "Newest First", "Oldest First"};
        AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
        builder.setTitle("Sort by");
        SortType currentType;
        switch (selectedFolder.getFolderOrigin()){
            case ONLINE:
                currentType = ApplicationPreferenceManager.getInstance().getOnlineFolderSortType();
                break;
            case LOCAL:
                currentType = ApplicationPreferenceManager.getInstance().getOfflineFolderSortType();
                break;
            default:
                currentType = SortType.NAME_ASC;
        }
        int checkedItem = -1;
        switch (currentType){
            case NAME_ASC:
                checkedItem = 0;
                break;
            case NAME_DESC:
                checkedItem = 1;
                break;
            case NEWEST_FIRST:
                checkedItem = 2;
                break;
            case OLDEST_FIRST:
                checkedItem = 3;
        }
        builder.setSingleChoiceItems(items, checkedItem, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                String result = items[which].toString();
                SortType newSortType = SortType.NAME_ASC;
                switch (result){
                    case "Name A-Z":
                        newSortType = SortType.NAME_ASC;
                        break;
                    case "Name Z-A":
                        newSortType = SortType.NAME_DESC;
                        break;
                    case "Newest First":
                        newSortType = SortType.NEWEST_FIRST;
                        break;
                    case "Oldest First":
                        newSortType = SortType.OLDEST_FIRST;
                        break;

                }
                enhancedAdapter.sort(newSortType);
                if (selectedFolder.getFolderOrigin() == FolderOrigin.LOCAL)
                    ApplicationPreferenceManager.getInstance().setOfflineFolderSortType(newSortType);
                else
                    ApplicationPreferenceManager.getInstance().setOnlineFolderSortType(newSortType);
                selectedFolder.sortFiles(newSortType);
                dialog.dismiss();
            }
        });
        AlertDialog alert = builder.create();
        //display dialog box
        alert.show();
    }

    private void getFolderFiles(Context context, EnhancedFolderViewerViewModel viewModel){
            if(enhancedAdapter.getItemCount() == 0) {
                backgroundThreadPoster.post(() -> {
                    try {
                        selectedFolder.getDataSource().getFilesFromDataSource(context, new IFolderDataSource.FolderDataSourceCallback() {
                            @Override
                            public void FolderFilesRetrieved(List<IDisplayFile> files, Exception exception) {
                                if (files != null) {
                                    //itemList = (ArrayList<EnhancedFile>) files;
                                    SortType initialSort;
                                    if (selectedFolder.getFolderOrigin() == FolderOrigin.LOCAL)
                                        initialSort = ApplicationPreferenceManager.getInstance().getOfflineFolderSortType();
                                    else
                                        initialSort = ApplicationPreferenceManager.getInstance().getOnlineFolderSortType();
                                    selectedFolder.sortFiles(initialSort);
                                    ArrayList<IDisplayFile> itemList = (ArrayList<IDisplayFile>) selectedFolder.getFiles();
                                    uiThreadPoster.post(() -> {
                                        viewModel.getFiles().setValue(itemList);
                                        recyclerView.hideShimmerAdapter();
                                    });

                                    //adapter = new FileGridAdapter(context,itemList);
                                    //gridview.setAdapter(adapter);

                                }
                                if (exception != null) {
                                    uiThreadPoster.post(() -> {
                                        Navigation.findNavController(binding.getRoot()).popBackStack();
                                        NotificationManager.getInstance().showSnackbar("Unable to retrieve files for folder " + selectedFolder.getName(), Snackbar.LENGTH_SHORT);
                                        recyclerView.hideShimmerAdapter();
                                    });

                                }
                            }
                        }, startingSortType);
                    } catch (MalformedURLException e) {
                        throw new RuntimeException(e);
                    }
                });
            }


    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        if(enhancedAdapter.getItemCount()>0)
            viewModel.getState().set("FileList",getGson().toJson(enhancedAdapter.getFiles()));
            //outState.putString("FileList",getGson().toJson(enhancedAdapter.getFiles()));
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onViewStateRestored(@Nullable Bundle savedInstanceState) {
        super.onViewStateRestored(savedInstanceState);
    }

    private void setTitle(String title){
        ((EnhancedMainMenuActivity)requireActivity()).setActionBarTitle(title);
        //((EnhancedMainMenuActivity)requireActivity()).overrideActionBarTitle(title);
    }
}

File: app/src/main/java/com/quigglesproductions/secureimageviewer/ui/enhancedfolderviewer/EnhancedFolderViewerFragment.kt
package com.quigglesproductions.secureimageviewer.ui.enhancedfolderviewer

import android.os.Bundle
import android.view.ContextMenu
import android.view.ContextMenu.ContextMenuInfo
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView.AdapterContextMenuInfo
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.lifecycleScope
import androidx.navigation.NavDirections
import androidx.navigation.Navigation.findNavController
import androidx.recyclerview.widget.GridLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.quigglesproductions.secureimageviewer.R
import com.quigglesproductions.secureimageviewer.databinding.FragmentFolderViewBinding
import com.quigglesproductions.secureimageviewer.managers.ApplicationPreferenceManager
import com.quigglesproductions.secureimageviewer.managers.FolderManager
import com.quigglesproductions.secureimageviewer.models.enhanced.file.IDisplayFile
import com.quigglesproductions.secureimageviewer.models.enhanced.folder.IDisplayFolder
import com.quigglesproductions.secureimageviewer.room.databases.unified.entity.RoomUnifiedFolder
import com.quigglesproductions.secureimageviewer.room.enums.FileSortType
import com.quigglesproductions.secureimageviewer.ui.EnhancedMainMenuActivity
import com.quigglesproductions.secureimageviewer.ui.SecureFragment
import com.quigglesproductions.secureimageviewer.ui.enhancedfolderlist.EnhancedFolderListViewModel
import com.techyourchance.threadposter.BackgroundThreadPoster
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class EnhancedFolderViewerFragment : SecureFragment() {
    //GridView gridview;
    private lateinit var binding: FragmentFolderViewBinding
    lateinit var selectedFolder: IDisplayFolder
    private lateinit var startingSortType: FileSortType
    //private val viewModel: EnhancedFolderViewerViewModelKt by viewModels()
    private val viewModel by activityViewModels<EnhancedFolderViewerViewModelKt>()
    private val folderListViewModel by activityViewModels<EnhancedFolderListViewModel>()
    private val backgroundThreadPoster = BackgroundThreadPoster()
    private lateinit var root: View
    private lateinit var swipeRefreshLayout: SwipeRefreshLayout

    @Inject
    lateinit var adapter: FolderFilesListAdapter
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        setHasOptionsMenu(true)
        binding = FragmentFolderViewBinding.inflate(inflater, container, false)
        root = binding.root
        swipeRefreshLayout = binding.folderViewerSwipeContainer
        val columnCount = resources.getInteger(R.integer.column_count_filelist)
        val layoutManager = GridLayoutManager(context, columnCount)
        binding.fileRecyclerview.layoutManager = layoutManager
        //binding!!.fileRecyclerview.showShimmerAdapter()
        selectedFolder = FolderManager.instance.currentFolder!!
        viewModel.folder.value = selectedFolder
        viewModel.folderListType.value = folderListViewModel.folderListType.value
        //restoreInstanceState(savedInstanceState)

        //registerForContextMenu(recyclerView!!)
        viewModel.fileSortType.value = when (selectedFolder.folderOrigin) {
            FolderOrigin.ONLINE -> ApplicationPreferenceManager.getInstance().onlineFolderSortType
            FolderOrigin.LOCAL, FolderOrigin.ROOM -> ApplicationPreferenceManager.getInstance().offlineFolderSortType
            else -> FileSortType.NAME_ASC
        }
        return root
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        val selectedFile = adapter.peek(item.itemId)
        when (item.groupId) {
            CONTEXTMENU_INFO -> {
                //new ItemInfoDialog(adapter.getItem(item.getItemId())).show(getSupportFragmentManager(),ItemInfoDialog.TAG);
                val bottomSheetDialog = BottomSheetDialog(requireContext())
                bottomSheetDialog.setContentView(R.layout.bottomdialog_fileinfo)
                val itemNameText = bottomSheetDialog.findViewById<TextView>(R.id.item_name)
                val folderNameText = bottomSheetDialog.findViewById<TextView>(R.id.folder_name)
                val artistNameText = bottomSheetDialog.findViewById<TextView>(R.id.artist_name)
                val catagoriesText = bottomSheetDialog.findViewById<TextView>(R.id.catagories)
                val subjectsText = bottomSheetDialog.findViewById<TextView>(R.id.subjects)
                selectedFile!!.dataSource.getFileMetadata(requiresRequestManager()) { metadata, exception -> //selectedFile.metadata = metadata;
                    itemNameText!!.text = selectedFile.name
                    folderNameText!!.text = selectedFolder.name
                    artistNameText!!.text = selectedFile.artistName
                    catagoriesText!!.text = selectedFile.catagoryListString
                    subjectsText!!.text = selectedFile.subjectListString
                    bottomSheetDialog.create()
                    bottomSheetDialog.show()
                }
            }

            CONTEXTMENU_SET_THUMBNAIL -> {
                /*val file = enhancedAdapter!![item.itemId]
                if (file is FileWithMetadata) {
                    backgroundThreadPoster.post {
                        modularFileDatabase.folderDao().setThumbnail(
                            selectedFolder as RoomEmbeddedFolder?,
                            file as RoomEmbeddedFile
                        )
                    }
                }*/
            }

            CONTEXTMENU_UPLOAD -> {}
        }
        return true
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel.fileAdapter = adapter
        setTitle(selectedFolder!!.name)
        //restoreInstanceState(savedInstanceState)
        //getFolderFiles(context, viewModel)
        initView()
        collectUiState()
    }

    private fun initView() {
        //adapter = FolderFilesListAdapter(context)
        binding.fileRecyclerview.adapter = adapter
        adapter.setOnClickListener(object : FolderFilesListOnClickListener {
            override fun onClick(position: Int) {
                viewModel.selectedFile.value = adapter.peek(position)
                viewModel.files.value = adapter.snapshot().items
                val action: NavDirections =
                    EnhancedFolderViewerFragmentDirections.actionEnhancedFolderViewerFragmentToEnhancedFileViewFragment(
                        position
                    )
                findNavController(root).navigate(action)
            }

            override fun onCreateContextMenu(
                menu: ContextMenu?,
                v: View?,
                menuInfo: ContextMenuInfo?
            ) {
                menu!!.setHeaderTitle("Options")
                val cmi = menuInfo as AdapterContextMenuInfo
                //EnhancedFile selectedFile = enhancedAdapter.get(position);
                val selectedFile: IDisplayFile? = adapter.peek(cmi.position)
                menu!!.add(CONTEXTMENU_INFO, cmi.position, 0, "Info")
                if (selectedFolder.folderOrigin == FolderOrigin.LOCAL) menu.add(
                    CONTEXTMENU_SET_THUMBNAIL,
                    cmi.position,
                    0,
                    "Set as Thumbnail"
                )
            }
        })
        swipeRefreshLayout.setOnRefreshListener {
            adapter.refresh()
            swipeRefreshLayout.isRefreshing = false
        }
    }

    private fun collectUiState() {
        //backgroundThreadPoster.post {
            viewLifecycleOwner.lifecycleScope.launch {
                viewModel.getFiles(viewModel.folder.value!! as RoomUnifiedFolder).collect { files ->
                    adapter.submitData(viewLifecycleOwner.lifecycle,files)
                }
            }
        //}

    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_folderview_fragment, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.folderview_fragment_sort -> showSortDialog()
            else -> return false
        }
        return true
    }

    private fun showSortDialog() {
        val items = arrayOf<CharSequence>("Name A-Z", "Name Z-A", "Newest First", "Oldest First")
        val builder = AlertDialog.Builder(
            requireContext()
        )
        builder.setTitle("Sort by")
        val currentType: FileSortType
        currentType = when (selectedFolder!!.folderOrigin) {
            FolderOrigin.ONLINE -> ApplicationPreferenceManager.getInstance().onlineFolderSortType
            FolderOrigin.LOCAL -> ApplicationPreferenceManager.getInstance().offlineFolderSortType
            else -> FileSortType.NAME_ASC
        }
        var checkedItem = -1
        checkedItem = when (currentType) {
            FileSortType.NAME_ASC -> 0
            FileSortType.NAME_DESC -> 1
            FileSortType.NEWEST_FIRST -> 2
            FileSortType.OLDEST_FIRST -> 3
        }
        builder.setSingleChoiceItems(items, checkedItem) { dialog, which ->
            val result = items[which].toString()
            var newSortType = FileSortType.NAME_ASC
            when (result) {
                "Name A-Z" -> newSortType = FileSortType.NAME_ASC
                "Name Z-A" -> newSortType = FileSortType.NAME_DESC
                "Newest First" -> newSortType = FileSortType.NEWEST_FIRST
                "Oldest First" -> newSortType = FileSortType.OLDEST_FIRST
            }
            //enhancedAdapter!!.sort(newSortType)
            if (selectedFolder.folderOrigin == FolderOrigin.LOCAL)
                ApplicationPreferenceManager.getInstance().offlineFolderSortType = newSortType
            else
                ApplicationPreferenceManager.getInstance().onlineFolderSortType = newSortType
            viewModel.fileSortType.update { newSortType }
            dialog.dismiss()
        }
        val alert = builder.create()
        //display dialog box
        alert.show()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
    }

    override fun onViewStateRestored(savedInstanceState: Bundle?) {
        super.onViewStateRestored(savedInstanceState)
    }

    private fun setTitle(title: String) {
        (requireActivity() as EnhancedMainMenuActivity).setActionBarTitle(title)
        //((EnhancedMainMenuActivity)requireActivity()).overrideActionBarTitle(title);
    }

    companion object {
        private const val CONTEXTMENU_INFO = 0
        private const val CONTEXTMENU_SET_THUMBNAIL = 1
        private const val CONTEXTMENU_UPLOAD = 2
        private const val LIST_UPDATE_TRIGGER_THRESHOLD = 75
    }
}
